Feature design / rfc: Formation flying

petervdmeer opened the pull request:
I combined the different formation flying features descriptions into a single rfc/feature design Markdown document to allow for easier and more efficient review of this feature.

Here is the [link to the formatted document](https://github.com/petervdmeer/endless-sky/blob/feature_formation_flying/docs/rfcs/text/0052-formation-flying.md) if you want Github to format the text for you (but please post any review-comments here).

petervdmeer committed:
Feature formation flying

Combine the different descriptions of the formation flying feature into a single rfc/feature design.

Signed-off-by: Peter van der Meer <peter.vd.meer@gmail.com>

petervdmeer committed:
Improved some texts.

petervdmeer committed:
Some more textual improvements

petervdmeer committed:
Add description of defining formations in fleets.

Note: proposal differs from current implementation in
https://github.com/endless-sky/endless-sky/pull/4471
This proposal should allow to set different formations within a single fleet.

petervdmeer committed:
Improved readability and added some options to the RFC

petervdmeer committed:
Fix some text display issues.

petervdmeer committed:
Removed non-needed line.

petervdmeer committed:
Minor readability improvements

kikotheexile commented:
This is honestly brilliant in concept and feels like a feature i would want out of most games i play, but i feel like its functionality is extremely limited in ES. 

Formations and postures like this assume same top speed, acceleration, and maneuverability for such quick in-and-out jumps and engagements we see in ES.  So although its rad as fuck, its not getting the utility it needs from the game environment itself. 

An amazing idea, a massive fish in a tiny pond, but systems need expanded to really get the full potential of this. 

I still love it and want to see it finished tho.

MCOfficer reviewed (approved):
Not commenting on the implementation/data format, just the feature.

I largely agree with kiko, for largely the same reasons. However, a quick glance at the "desired functionality" section tells you exactly why this is a great feature. It's a much needed first step, even though it won't bring a lot of immediate benefit.

Zitchas reviewed (approved):
I haven't worked out all the math for the patterning, but this looks like all the sort of information I'd need to play with building my own fleets. 

This is definitely something that I think will add a lot of detail and interest to the game.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+- Ships automatically filling up positions of killed ships (by moving to the earliest formation-position available to the ship).
+- Ships in a formation will sync based on the movement vector of the reference/lead ship (not the facing vector of the reference/lead ship).
+
+## Highly desired but currently not implemented functionaly
```
We should be prescriptive with RFC language and content - the goal is to outline how the thing will work. A detailed accounting of things that don't work yet doesn't quite belong in a description of how things will work.

petervdmeer replied:
We also want the things that don't work yet to be working in the future. But I just changed the text to a description to "how things should be working" without much reference to what is working today.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+## Flying behaviour
+
+### Formation turning
+The positions in a formation as implemented in [ES-#4471](https://github.com/endless-sky/endless-sky/pull/4471) exclusively depend on the position and movement direction of the lead ship.
```
I would expect this section to describe the goal for turning. If there are multiple desired-yet-exclusive behaviors, this section should identify each of them and identify the associated data keyword used to select the behavior.

petervdmeer replied:
There are multiple desired behaviours, but they are not mutually exclusive. The keywords for symmetry are described, I don't see a need for other keywords after the update I just did.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+
+### Formation positions
+
+Actual formation positions are determined by the order in which ships are in the ships-list and only applies to ships in the system. If formation-flying ships from outside the system join, then they take their relative positions causing other ships already in formation to move to their new relative positions.
```
This sounds like a current implementation detail. Is the desired goal to have ships behave in this way?

petervdmeer replied:
If we have different types of ships in a single formation (for example Heavy Freighters and Interceptors), then this would allow the player some controls to ensure that the Heavy Freighters are more to the inside of the formation while the Interceptors would go more to the outside of the formation.

The ring-settings however also allow for a similar type of control, so we could also choose that we want "first come - first position" formations and handle what goes to the inside and what goes to the outside only through the ring-settings.

Implementation wise it doesn't make much difference. Both are quite straight-forward to implement.

petervdmeer replied:
I just added an example of the ring feature in https://github.com/endless-sky/endless-sky/pull/4471 to give an impression of how that part should work.

petervdmeer replied:
Ring settings are now explicitly mentioned in this section as well.

tehhowch commented on docs/rfcs/formation_flying.md:
````diff
@@ -0,0 +1,159 @@
+```
+formation <name>:
+	symmetry ["transverse"] ["longitudinal"]
+		rotational <nr>
````
Please use the [same syntax](https://github.com/endless-sky/endless-sky/wiki/DataFormat#grammar-specifications) as the wiki for identifying data types

petervdmeer replied:
Should be better now after my last update.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+
+Meaning of the keywords:
+- `formation <name>`: Begins a new definition for a formation with name `<name>`.
+- `line <x> <y> <slots> <direction>`: Begins a line at position x,y for a line with `<slots>` positions for ships. The very first ship in the line is at exactly x,y, the line stretches into the direction of `<direction>` with ships `<spacing>` distance apart. The last ship in the line is at `<slot> * <spacing>` distance from the first.
```
You should include examples in these sections. For example, what are the prescribed values of a direction? Up? Forward? North?

Also, does it make sense to perhaps support a syntax like
```
line [<shipCount#>]
  start <x1#> <y1#> # defaults to (0, 0)
  end <x2#> <y2#>
```

petervdmeer replied:
Yes, I was doubting about supporting multiple definition formations, but this makes a lot of sense.
I might also include polar coordinates as possible method to specify coordinates.

petervdmeer replied:
Defaults and directions should now be in the specification. Polar coordinates as well.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+- `spacing <nr>`: The amount of space between ships on a line. This number is in "ship sizes", so 2 is twice the `max(width,lenght)` of a ship.
+- `repeat <deltaX> <deltaY>`: The location where to repeat the line if the formation needs to grow (relative to x,y of previous repeated line).
+- `increase <nr>`: The amount of extra slots to add to the line on each grow step.
+- `arc <refX> <refY> <radius> <angle> <segment>`: Specify partial or full circle segment. `<refX>` and `<refY>` specify the anchor/reference point for the arc.
```
Use `center` instead of `ref` - I think it is more commonly associated with your intended use in mathematics / algebra / geometry.

Also, does it make sense to use named parameters e.g.
```
arc 0 0 100 -45 45
``` 
vs
```
arc 100
  center 0 0 # default 0 0
  degrees -45 45 # default 0 360
```

petervdmeer replied:
Yes, center is more appropriate. And named parameters looks a lot more readable indeed.

petervdmeer replied:
I changed to `anchor`, which I think is commonly used in other programs. If you prefer center then I can change it to center.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+- `increase <nr>`: The amount of extra slots to add to the line on each grow step.
+- `arc <refX> <refY> <radius> <angle> <segment>`: Specify partial or full circle segment. `<refX>` and `<refY>` specify the anchor/reference point for the arc.
+   - `<radius>` specifies the radius for the arc.
+   - `<angle>` specifies the direction to the start point of the arc angle (0 to 360)
```
What sign convention does this use? I think for ease of use we should adopt standard mathematical convention - the game implementation can handle the conversion to graphics convention.

petervdmeer replied:
I will add some specification here on the units and conventions used for those numbers.

petervdmeer replied:
For ease of use I for now used the Endless Sky convention where possible.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+- `arc <refX> <refY> <radius> <angle> <segment>`: Specify partial or full circle segment. `<refX>` and `<refY>` specify the anchor/reference point for the arc.
+   - `<radius>` specifies the radius for the arc.
+   - `<angle>` specifies the direction to the start point of the arc angle (0 to 360)
+   - `<segment>` (-360 to 360) specifies how much of a circle is used; 0 is single formation position, 90 is a quarter circle, 360 is a full circle, an positive segment is clockwise, a negative segment is counter-clockwise.
```
I think degrees is a better label

petervdmeer replied:
Degrees is an unit of measurement, not a measured property itself. I would prefer a keyword that describes the measured property, not the unit in which it is measured.

tehhowch replied:
I'm not sure what the more appropriate usage guideline label for the angular measure of an arclength would be. "segment" just doesn't seem to fit when the intent is to convey something like "this is an arc centered on (X, Y) with a radius of 100 that starts at an angle of 30° and spans 90°"
Having written that, I think "`<span#>`" makes more sense than "`<segment#>`".

I also am not sure if it makes sense to use the syntax here, where we define the starting angle and the angular span of the arc: specifying the starting and ending angles of the arc seems just as valid (and unfortunately has the same signature of "number number" if done on a single line).

petervdmeer replied:
This information is now (embedded) in the start and the end keywords/tags.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+   - `<radius>` specifies the radius for the arc.
+   - `<angle>` specifies the direction to the start point of the arc angle (0 to 360)
+   - `<segment>` (-360 to 360) specifies how much of a circle is used; 0 is single formation position, 90 is a quarter circle, 360 is a full circle, an positive segment is clockwise, a negative segment is counter-clockwise.
+   - `"nofirst"`: Don't fill the first position (starting position) of the arc. Usefull if another line or arc fills that position.
```
Maybe these should be child nodes, "skip", taking a number or the keywords "start" and "end". This child node could also work for lines then, where the specific slot number to skip can be easily computed.

petervdmeer replied:
Those keywords have been completely removed.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+   - `<segment>` (-360 to 360) specifies how much of a circle is used; 0 is single formation position, 90 is a quarter circle, 360 is a full circle, an positive segment is clockwise, a negative segment is counter-clockwise.
+   - `"nofirst"`: Don't fill the first position (starting position) of the arc. Usefull if another line or arc fills that position.
+   - `"nolast"`: Don't fill the last position (end position) of the arc. Usefull if another line or arc fills that position.
+- (arc) `spacing <nr>`: Specifies the minimum amount of spacing between ships on the arc. Ship positions are evenly spaced among the arc, with at least the minimum as given here.
```
Should spacing should be interpreted the same for both geometry elements? It seems like there is differing behavior for arcs and lines, due to the different styles of start and end specification.

petervdmeer replied:
Both those differences comes from the difference in the "repeat" behaviour;
- for lines you can calculate upfront how many slots there are in each repeated line (and specify all line/slots information upfront).
- I don't know a simple specification to do the same for the arc sections; as sections grow there will be more and more slots to add more ships, but not in a linear sense as with lines.

The difference in start and end specification is also due to the "repeat" behaviour difference. For lines I always know where it starts and ends and if I want to skip a start or end slot, then I can just make the line 1 slot shorter (and calculate the new start position if it was the start I was changing).
For arcs I cannot determine the exact ship slot positions of the follow-up/repeated lines with a simple definition. And to be able to connect 2 arc sections I need to be able to perform skip on one of them (to avoid having 2 ships on the same formation position).
Alternatives are to accept gaps between the sections, accept overlapping positions, automatically detect overlapping positions or to always leave out the start and the end positions for arcs.
Always leaving out the start and end positions might be better (to keep the specification simpler and the amount of exceptions smaller), I'll update the specification for that.

petervdmeer replied:
I updated the specification to be more explicit about which keyword applies to lines and which to arcs and how they apply differently.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,159 @@
+   - `"nofirst"`: Don't fill the first position (starting position) of the arc. Usefull if another line or arc fills that position.
+   - `"nolast"`: Don't fill the last position (end position) of the arc. Usefull if another line or arc fills that position.
+- (arc) `spacing <nr>`: Specifies the minimum amount of spacing between ships on the arc. Ship positions are evenly spaced among the arc, with at least the minimum as given here.
+- (arc) `repeat <deltaRadius>`: Specifies the repeat vector among which this arc is repeated.
```
A vector requires two parameters.

I'm also of the opinion that we should try to make a repeat node be interpreted similarly for arcs and lines, i.e. have the same child nodes

petervdmeer replied:
Each repeated arc uses the same center, the same direction for start, the same direction for end but a different radius. Maybe I should change the word vector here.

petervdmeer replied:
I made some changes to make the keywords more similar. The deltaRadius is now embedded in the start (and the end) tags.

tehhowch reviewed (changes requested):
Thanks for writing up the RFC! The biggest request here is to basically revise it all, to use a more RFC-like language. The idea behind these is that, upon adopting one, we have a definite guideline / specification. It is good to acknowledge the departure from current behavior, but divulging implementation details isn't appropriate unless we are debating between implementations.

I didn't get to the latter part of the RFC, and will hold off on reviewing it until after an update.

toilethinges reviewed (approved):
This looks really cool! Without looking at the code, I definitely think this is a great feature and one that I think would be really interesting to have in the game.

petervdmeer committed:
Update to use grammar specifications.

petervdmeer committed:
Update and added some keywords to add some features.

petervdmeer committed:
Rewritten from informational style to specification style.

petervdmeer commented:
Document is largely rewritten (as requested by @tehhowch); this RFC now describes how formations should work without any reference to what is currently working and what not.

I still have some review-comments to process.

petervdmeer committed:
Made the keywords for line and arc more similar.

The concepts behind line and arc are a little bit different, but tried to use the same keywords where possible.

petervdmeer committed:
Split the document and keyword section into two parts; defining and usage

petervdmeer committed:
Performed some textual improvements. Removed some duplicate texts.

petervdmeer committed:
Clarified clockwise remark.

petervdmeer committed:
Add some sensible default values.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+# Formation definitions
+## Signs and axis units for formation coordinates
+- Coordinates are relative to the center of the formation(pattern), with coordinate 0,0 being the center.
+- Coordinate signs (plus/minus) in formation definitions follow the default Endless Sky convention;
```
If it is helpful, you can upload a file (to `rfcs/imgs/<descriptive-name>`) and embed it in this document.

petervdmeer replied:
Yes, but this could also be a "less is more" situation. If I look at the `"engine"` keyword in https://github.com/endless-sky/endless-sky/wiki/CreatingShips, then it just says;

**"engine": the (x, y) coordinates, relative to the center of the sprite, where engine flares should appear. Positive y is up; negative y is down"**

This describes with a lot less words how the ships coordinate work (x is not specified at all, probably because it works as everybody expects).

petervdmeer replied:
I tried to simplify the text a bit based on this example.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+- Polar coordinates in formation definitions start with an angle and then have a distance.
+   - Angle 0 is towards the front of the formation.
+   - Angle 90 is towards the right of the formation.
+- Coordinate axis lengths (both for carthesian as well as distances in polar coordinates) in formation definitions are measured in ship sizes, where 1 corresponds to the radius of the largest ship actively participating in a formation.
```
A downside of this is that you cannot define a formation with a particular antimissile range unless you also limit which ships can participate in the formation. However, I think it would be possible to use a keyword when referring to the formation pattern that determines where the pattern gets the scale magnitude. For example, you might want one instance of a Delta formation to treat the unit length as 400 px, and a different instance to treat the unit length as the largest sprite height or width of participating ships.

petervdmeer replied:
That is true. If you want to define formations using very exact positions in px, then you also need to be able to set an exact unit length (or only use one type of ship in the formation).

If we add the keyword for scaling to the formation definition, then we would get 'auto-scaling' formations and 'exact position' formations. Those formations types behave quite differently, so I think that that is nicer than adding it to the sections where we attach a formation to a fleet or npc. Do you agree?

petervdmeer replied:
I added px-based formations as default and also allow size-based formations as option.

tehhowch commented on docs/rfcs/formation_flying.md:
````diff
@@ -0,0 +1,220 @@
+The basic structure of a definition of a formation in the data-files would look like:
+```
+formation <name>:
+	symmetry [transverse] [longitudinal]
````
It might be worth just using "x" or "y" instead of the more formal names.

petervdmeer replied:
Yes, that might just fit better. Changed accordingly (and used the formal names in the description).

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+- Positions in a formation are sequentially assigned.
+   - Actual formation positions are determined by the order in which ships are in the ships-list
+      - If new ships join a formation, then they take their relative positions causing other ships already in formation to move to their new later relative positions.
+      - The use of the ships-list order allows players some control over which ships appear where in the formations.
```
This doesn't seem to work with the previous statements re: symmetry. If the front can become the back, or the left the right, or if we have e.g. hexagonal rotational symmetry, doesn't that mean we can lose the ordering intentions?

petervdmeer replied:
Good catch!

When I was thinking about the keywords for symmetry, then I was mostly thinking about the protective formations that have the flagship in the center. The current "center oriented" formations are ordered "inside to outside"; so if you have the ships ordered in a way that all your heavy transport ships are on the inside of the formation, then they will still be on the inside of the formation when the formation is rotated, flipped or mirrored.

Most formations that are "tailing" are ordered "front to back" and most formations that are "ahead" are ordered "inside to outside". Those can be flipped along the y-axis and still have the same type of ships at the front and center of the formation. I never considered to give them other symmetry settings.

I added some text to the RFC that symmetry is only for symmetry properties around the formation center.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+
+Meaning of the keywords
+- `formation <name>`: Name of the formation to form (around the lead-ship)
+- `class <shipclass>`: Indicates for fleets/governments that the given formation should only be applied to the given class of ships. (TODO: attributes filter instead of class?)
```
I would suggest using a keyword that introduces a locationfilter definition, which would allow you to select applicable ships just like you would for an assisting or boarding mission

petervdmeer replied:
Yes, a location filter is the appropriate filter, it can select on things like attributes, categories and on outfits.

petervdmeer replied:
Changed to locationfilter yesterday.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+Meaning of the keywords
+- `formation <name>`: Name of the formation to form (around the lead-ship)
+- `class <shipclass>`: Indicates for fleets/governments that the given formation should only be applied to the given class of ships. (TODO: attributes filter instead of class?)
+- `ring <nr#>`: Area in the formation where the ship should appear (higher numbers is more outwards on most formations).
```
I thought the order in the formation was based on the ship list ordering?

petervdmeer replied:
There are 2 mechanisms described in this RFC to determine formation positions;
The ring mechanism overrides the ships-list-ordering. Ships with a higher ring are always (later/more outside/more back/more to the end) of the formation and rings can also ensure that space in (the beginning/more inside/more to to front/more to the begin) of the formation is left empty.

Here is the example from #4471 where the center of a delta formation is left empty by using a ring setting for the ships in the delta-formation: ![formations_ring_example](https://user-images.githubusercontent.com/35403542/79698287-50bdca80-8288-11ea-86ce-aaf65ecef3a7.png)

thomasballinger replied:
I'm confused about these rings vs rings in the old code because the rings in this image have different shapes. In the code a ring is just the next repeat of each line. Are these two different kinds of rings?

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+- `reference point [<x#> <y#>] [<direction#>]`: Form the formation around a specific fixed point in the system (direction indicates the formation direction). If x and y are not given, then use system center (point 0,0).
+- `reference ship <name>`: Form the formation around the ship with name `<name>`.
+- `reference ships <class>`: Form formation around the ships with class `<class>`. Can be given multiple times, for example for `Transport` and `Heavy Freighter` to form protective formations around transportships.
+- `reference mission`: Form formation around mission (NPC) ships. Ships using this reference are evenly distributed around mission ships.
```
There are a lot of missions that could be active at once - does this take a mission name, e.g. so NPCs from an optional side quest know which other mission's ships they're supposed to protect?

petervdmeer replied:
This setting was actually intended for players that fly around with active missions; form a formation around all the NPC ships that accompany the player.

But I decided to remove this keyword completely now that I thought a bit more about this; there can be many different missions active at any time:
- Some NPCs that follow the player can be hostile. Forming a formation around then is not a good idea.
- Some NPCs are fighters/fleets that join the player into battle. Forming a formation around friendly fighters and warships would also be odd.
- Some NPCs that follow the player have formations of their own. The interaction between the formations could become complex and unpredictable.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+- `reference ships <class>`: Form formation around the ships with class `<class>`. Can be given multiple times, for example for `Transport` and `Heavy Freighter` to form protective formations around transportships.
+- `reference mission`: Form formation around mission (NPC) ships. Ships using this reference are evenly distributed around mission ships.
+- `reference planet <name>`: Form formation around planet with name `<name>`. Not to be used by players, but can be usefull for defense fleets around a planet. Note that planets need to have a `direction` to make this option work nicely. Direction can be random (easy) or based on orbit around the star (a little bit more complex).
+- `reference planets <nr1#> [<nr2#>]`: Form formation around the `<nr1>`th planet in the system. If two numbers are given, then form formations around all planets in the range `<nr1>-<nr2>`. If `<nr2>` is -1, then form formations around the `<nr1>`th planet up to the last planet in the system.
```
I don't think this syntax is particularly useful, and it's probably prone to issues when events modify systems. In general, you cannot assume that map.txt is the state of a player's map at any given time

petervdmeer replied:
Agreed, I will remove this one.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+- `reference mission`: Form formation around mission (NPC) ships. Ships using this reference are evenly distributed around mission ships.
+- `reference planet <name>`: Form formation around planet with name `<name>`. Not to be used by players, but can be usefull for defense fleets around a planet. Note that planets need to have a `direction` to make this option work nicely. Direction can be random (easy) or based on orbit around the star (a little bit more complex).
+- `reference planets <nr1#> [<nr2#>]`: Form formation around the `<nr1>`th planet in the system. If two numbers are given, then form formations around all planets in the range `<nr1>-<nr2>`. If `<nr2>` is -1, then form formations around the `<nr1>`th planet up to the last planet in the system.
+- `reference wormholes <nr1#> [<nr2#>]`: Form formation around the `<nr1>`th wormhole in the system, similar to planets (with ranges). This allows for NPC pirates to form an ambush around a wormhole.
```
Wormholes are always named planets and the system + planet combination is unique identifier

petervdmeer replied:
Changed to explicit names (with the system name being optional).

petervdmeer replied:
Also combined planet and wormhole into single object definition.

tehhowch commented on docs/rfcs/formation_flying.md:
````diff
@@ -0,0 +1,220 @@
+		ring <nr#>
+		reference <...> ...
+```
+This is used in savegames to store the formation used by the ship.
````
This seems extremely implementation dependent - I don't think we should prescribe in the RFC how formation information gets serialized on instantiated fleets.

I think the important thing to prescribe is the `fleet` and `npc` usage, as this is how a mission writer could use the feature.  
To me, a formation requires more than one ship, and so we should not explicitly allow setting the formation behavior at the ship level. (If the specific implementation makes that implicitly possible, that's fine, but it shouldn't be an advertised feature.)

petervdmeer replied:
I removed the ship keyword. It is indeed not relevant for mission writers, but what we write here will also be relevant for programmers that implement the feature.

tehhowch commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,220 @@
+formation <name>:
+	symmetry [transverse] [longitudinal]
+		rotational <nr#>
+	line <direction#>
```
There isn't a description for the required parameter below

petervdmeer replied:
Sorry, the keyword was no longer used here (replaced by `angle` on the start keyword).

tehhowch reviewed (commented):
I have more comments but not the time to formulate them, so here's some more feedback.

petervdmeer committed:
Processed some (symmetry and keyword) review comments

- Added some additional symmetry information.
- Removed "reference mission" completely.
- Removed keywords that were no longer used.
- Switched o using only explicit names for planets and wormholes.

petervdmeer committed:
Remove ship keyword.

petervdmeer committed:
Changed to more informal symmetry axis names.

petervdmeer committed:
Add locationfilter criteria to select formation-member-ships and formation-lead-ships.

petervdmeer committed:
Changed planet and wormhole references to single object reference.

petervdmeer committed:
Tried to simplify the coordinate text according to review comment.

petervdmeer committed:
Handle some unhandled review comments.

Some review comments were not fully handled. This commit handles some of them.

petervdmeer committed:
Clarified the specification a bit.

Moved angle to its own keyword.
Split keywords over lines and arcs and specified meaning in their own sections.
Re-added the skip keyword (since explicit seems more clear there).

petervdmeer committed:
Allow formations to be applied multiple times

This allows the same formation to be used for small and large ships, without the large ships pushing the small ships far apart.
This would usually be used with the small and large ships in different rings.

petervdmeer committed:
Make pixel-based positions the default

And allow ship-radius, ship-width and ship-height based positions available as options.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+
+# Formation definitions
+## Signs and axis units for formation position coordinates.
+- (x,y) coordinates are relative to the center of the formation(pattern), with coordinate 0,0 being the center, with positive x being towards the right side of the formation and positive y being towards the front of the formation.
```
Is the center of the formation always a ship? (I'm pretty sure yes, but was confused about this when I first read it)

petervdmeer replied:
Yes, the current examples always have a ship as the center of the formation, but I created this spec (and the example implementation) with the intention of also allowing formations around things like planets and wormholes (with the planet or wormhole being the center of the formation).

petervdmeer replied:
Updated the text a bit (I hope it is more clear now).

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+
+Meaning of the keywords:
+- `formation <name>`: Begins a new definition for a formation with name `<name>`.
+- `symmetry [x] [y]`: Indicates if a formation should be considered symmetric. Symmetric formations allow for nicer and more desired behaviour on turns.
```
Another possible name for this is `allow-flip x` or `flippable x`; might reflect better what this does.

petervdmeer replied:
Keywords `flippable [x] [y]` and `rotatable <angle>` will indeed reflect better what those keywords do. I'll change the spec to use those keywords.

petervdmeer replied:
Changed to `flippable` and `rotatable`.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+- `symmetry [x] [y]`: Indicates if a formation should be considered symmetric. Symmetric formations allow for nicer and more desired behaviour on turns.
+   - `x`: Indicates that a formation can be mirrored along the x(/horizontal/transverse)-axis and still be the same formation.
+   - `y`: Indicates that a formation can be mirrored along the y(/vertical/longitudinal)-axis and still be the same formation.
+   - `rotational <nr#>`: Indicates rotational symmetry (nr is in range 0 to 360). Rotational symetric formations don't need to turn fully to become aligned with the leadship/reference again.
```
- I think I might want to use small n here, like 1. Will that be ok implementation-wise?
- Would using rotational symmetry order (3 for a triangle, 5 for a pentagon) might clearer here? I'm not sure it would, just an idea.

petervdmeer replied:
Using a small n like 1 would be fine implementation wise. This would result in a formation that doesn't turn a lot even if the lead ship does changes direction a lot.

I expect angles to be more clear, especially if we remove the keyword symmetry (and just use rotatable here).

petervdmeer replied:
I added some text about using a small n and when it could be desired to do so.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+   - Example: A delta (triangular) formation would also have longitudinal symmetry, it can be mirrored along the longitudinal axis and would still be roughly the same shape.
+   - Mirror, flip and rotate actions are all around the center of the formation, so only symmetric properties around the center should be considered.
+   - Example: A delta-tailing (triangle behind the flagship) formation should only be considered `transverse` symmetric.
+- `line`: Begins a line.
```
I sort of want a line that grows from the middle out. Not sure if it makes sense for this to be a property in here.

petervdmeer replied:
Yes, that makes a lot of sense and an earlier version of the formations spec (around the time of the initial proposal in https://github.com/endless-sky/endless-sky/issues/4438) actually had a keyword for this.
It also had a keyword for starting from the end, and a keyword to have repeat lines alternatively starting from the start and the end. And a keyword for extending lines so that positions/positionnumbers move along the line when the formations grows.

I removed all those keyword in later versions of the spec, since the spec was already quite complex and I was able to also create centered lines by having 2 opposite lines starting from about the same position.
The alternating start-from-begin and start-from-end were also used in the "delta tailing" formation without those keywords by having 2 different lines repeating interwoven.

Maybe we should add the `center` keyword back, since it will make creating a centered line quite a bit easier. Do you feel we should also add some of those other keywords back?

thomasballinger replied:
I agree that it's gotten quite complex. Unlike polar coordinates or the ability to use spacing or start/end to specify number, this is something I wanted when trying to build a formation, so I think we should add this one back. Good point about two lines going out though, if that works maybe not. Won't two lines in opposite directions have one line filled first, then the other? Or has that changed?
![2020-06-13 07 11 43](https://user-images.githubusercontent.com/458879/84570933-33b3ee00-ad45-11ea-9119-abf1ef4128b8.gif)

petervdmeer replied:
> Won't two lines in opposite directions have one line filled first, then the other? Or has that changed?

You are right. One line would be filled first and then the other. This is a good reason to (re-)add the keyword for centring. I'll add it back to the spec.

petervdmeer replied:
I added `centered` back as keyword and I added the keywords `originating` and `extending` to help proper centering over multiple rings.

I also added alternating. The alternating keyword will work nicely for the pattern you posted here if ships move to earlier positions.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+      - An angle can also be implicitly given by specifying a begin and end-point.
+      - Repeat lines are normally in the same direction as the original line, but if an angle is given for repeat lines, then this gives the additional direction change for repeat lines.
+   - `spacing <nr#>`: The amount of space between slots/ships on a line (by default in pixels, unless `radius`, `width` or `height` was given).
+      - The default spacing is 2; so twice the radius of the largest ship in the formation.
```
Should this say "The default spacing is radius 2?"

petervdmeer replied:
Yes, that is what it should have said before my last update. But my last update changed all defaults to be in pixels, so then the default spacing would also need to be in pixels. (160 pixels, twice the radius of a Sparrow was working nicely as default in some earlier tests).

I'll update this text.

petervdmeer replied:
Text was updated. And then deleted due to removal of this keyword.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+      - Repeat lines are normally in the same direction as the original line, but if an angle is given for repeat lines, then this gives the additional direction change for repeat lines.
+   - `spacing <nr#>`: The amount of space between slots/ships on a line (by default in pixels, unless `radius`, `width` or `height` was given).
+      - The default spacing is 2; so twice the radius of the largest ship in the formation.
+         - If a line has an end-coordinate, slots and no spacing, then spacing is automatically calculated (instead of the default being used).
```
Both an implementation and spec note: it maybe useful to validate formations are not over-constrained at load time or spit out warnings, whatever the pattern is for invalid data. These notes about what will happen in over-constrained situations are great, but if there's no possible need for some combinations of properties it might be nice to make them warn or be invalid.

petervdmeer replied:
The older versions of the spec didn't support the `end` keyword, because it overlaps with start-angle, slots and spacing. We could also completely remove one of the following combinations and still have the same "expressiveness" without multiple ways to specify the same information:
- the `spacing` and `start angle` keywords
- the `slots` and `start angle` keywords
- the `end` keyword

Most people I know seem to think of lines in "start" and "end" coordinates and specifying spacing is a bit odd, so it is probably most intuitive for people working with formation-patterns if we remove the spacing and start-angle keywords from the spec (and always calculate those automatically based on start-coordinate, end-coordinate and slots).

What would you prefer, remove some of those keywords? (And if so, then which one?) Or would you prefer have some load-time checks on over-constraining due to overlap in keywords?

thomasballinger replied:
I would prefer a check that both aren't used in the same place, but I'd like to keep slots.

I don't love using angle, but I do like specifying how many ships are in a row. I think plugin/content authors will want to describe exactly how a set of e.g. 7 ships in one fleet will fly, which makes start/end pretty finicky since they need to guess numbers that will make the right number of ships fit, and number of ships in a line could change when size of the ships changes.

petervdmeer replied:
My proposal was to keep 'start' (coordinate), 'end' (coordinate) and 'slots' and to remove the keywords 'angle' and 'spacing' for lines. That seems to match with the preference you indicate here.
Slots will be the only number that determines how many ships there will be on a line, so that should be quite straight-forward for plugin/content authors.

petervdmeer replied:
Maybe a bit fast, but I applied the changes as discussed. Let's see if this makes the specification simpler. (And if not, then let's discuss how to proceed.)

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+         - If a line has an end-coordinate, slots and no spacing, then spacing is automatically calculated (instead of the default being used).
+      - Ships are `<spacing#>` distance apart. The last ship in the line is at `<slot#> * <spacing#>` distance from the first.
+   - `slots <nr#>`: The amount of slots on a line. (Default is 1, meaning that the line is a single point.)
+      - Or the amount of slots to increase/decrease on each growth step when given in a line repeat section.
```
good change, I missed being able to use a negative step in a repeat in the old implementation

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+   - `repeat`: Section for repeating a line when the formation needs to grow.
+      - Repeat lines that reach a size of 0 will not repeat further.
+- `arc`: Begins a partial or full circle.
+   - Angle start and end positions are always interpreted clockwise from start to end (to avoid ambiguity).
```
Can I get the to fill in the opposite direction?

petervdmeer replied:
Good catch. Older versions of the spec allowed this by interpreting angles of 0 to 360 clockwise and angles of 0 to -360 counter-clockwise, but this was lost in the last version of the spec.
I think we should also check here if we want to remove some keywords to remove ambiguity, similar to the above suggestion for lines and we should also check that whatever we choose still supports filling in the opposite direction.

We might also want to add an `clockwise` and/or `counter clockwise` keyword specifically for arcs.

petervdmeer replied:
I fixed this (and the conflicting keywords) by removing the end-coordinate and spacing for arcs and by allowing the angle to range from -360 to 360. Please feel free to re-review the update and check if it is better or not.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+
+## Turning behaviour
+- The front of an active formation is aligned with the movement vector of the ship/object/planet/point that the formation is formed around.
+   - If the formation is around a lead ship that has a facing, then one third of a degree delta is added for every degree that the lead ship turns.
```
This might be standard ES terminology, so my apologies if this is just for my education, but I was confused be "a lead ship having a facing" meaning "a lead ship whose velocity is different than its heading".

petervdmeer replied:
This text is a bit confusing indeed. I'll add that the front (facing) of the ship is pointing in a different direction that where it is moving to.

thomasballinger commented on docs/rfcs/formation_flying.md:
```diff
@@ -0,0 +1,232 @@
+      - This is up to a maximum of 90 degrees for the lead-ship (and thus 30 degrees for the formation).
+   - If the formation is formed around a ship/object that is not moving, then the front is aligned with the facing direction of the lead-ship/object.
+      - Planets have an orbit specified, the facing direction could be calculated based on the orbit.
+   - If the lead ship/object is not moving and not having a facing direction, then the front is in the direction of the system center.
```
Do some ships not have a facing direction?

petervdmeer replied:
This is for formations that are around things like planets and wormholes, where the planet or wormhole is the center of the formation.
I'm not sure if all of those have a facing direction.

thomasballinger reviewed (commented):
This seems well thought out. The only thing I think I want is the ability to fill lines from the middle instead of from start to end.

thomasballinger commented:
One more thing I'm thinking about is "if these lines and arcs don't work for the formation I want to create, can I just use one line per ship to get the behavior I really want. I think the answer is yes.

petervdmeer commented:
> One more thing I'm thinking about is "if these lines and arcs don't work for the formation I want to create, can I just use one line per ship to get the behavior I really want. I think the answer is yes.

Yes, that is correct.
You could also create a formation(pattern) uniquely for each ship (with each formation being 1 line of 1 position), then you would also not have the automatic position replacement.

petervdmeer committed:
Fix minor typo.

petervdmeer committed:
Processed first part of review comments

- Changed symmetry to flippable and rotatable. Those properties can also be used on non-symmetric formations.
- Added example of rotatable with 1 degrees of less.
- Added some text to the end coordinate about the coordinate systems.
- Clarified the text about coordinate axises a bit.

petervdmeer committed:
Make sure all coordinates and lengths have the radius/width/height keywords

petervdmeer committed:
Remove the angle and spacing keywords for lines.

Based on review comments. Specifying lines with begin and endpoints and slots seems more natural.

petervdmeer committed:
Clarify the text about lead-ship facing/heading compensation.

petervdmeer committed:
Removed end-statement and spacing for arcs and allowed counter-clockwise

Also added slots to arcs (to compensate for removing the spacing).
And updated some texts to make the meaning of the keywords more clear.

petervdmeer committed:
Add centered keyword as discussed during review.

petervdmeer committed:
Add regulated keyword to control the spacing between slots on repeat lines.

This will be most useful for arcs, but can also be of use for lines.

petervdmeer committed:
Add alternating keyword

Also removed an occurence of spacing that should have been removed earlier.

petervdmeer committed:
Allow lines and arcs to have multiple repeat sections.

petervdmeer committed:
Added originating and extending keywords.

Those keywords are likely to be useful for formations that extend in 2 directions (like horizontal lines).

petervdmeer committed:
Fixed some defaults and added texts about multiple start/end/anchor keywords

petervdmeer committed:
Removed unused keyword from repeat-line.

thomasballinger reviewed (approved):
Thanks for addressing all this, I think this sounds great.

petervdmeer committed:
Changed radius keyword to diameter to match the expectation of ship-size better.

As discussed on Discord (in the context of this review).

petervdmeer committed:
Minor format fix for alternating keyword

petervdmeer committed:
Remove keywords originating and extending from the RFC.

Those keywords provided only a small amount of additional value and added a large amount of complexity.

petervdmeer committed:
Merge remote-tracking branch 'upstream_community/community' into feature_formation_flying

petervdmeer committed:
Update formation flying RFC to match the current/latest template.

petervdmeer committed:
Allow specifying single-points/slots as formation-positions.

This should make it easier for content-creators to create their own patterns.

MCOfficer commented:
CI failure is due to an outdated workflow. Either merge master or ignore it, RFCs arent testable anyways.

petervdmeer committed:
Merge branch 'community' into feature_formation_flying

petervdmeer committed:
Merge branch 'community' into feature_formation_flying

petervdmeer committed:
Extend unresolved questions with some more weaknesses in the current formations-spec.

petervdmeer committed:
Add new PR and replace point and slot by position.

Positions in a formation sounds better than points or slots in a formation.

petervdmeer committed:
Add category keyword to allow multi-category formations in a single definition.

petervdmeer commented:
I just updated the specification to use some less mathematical and more context-aware keywords.

petervdmeer committed:
Add suggestion on additional filtering.

petervdmeer committed:
Add filter keyword (and resolve open issue around it)

TheGiraffe3 commented:
This should be closed because if it’s in the main game it doesn’t need to be implemented twice here (also, a lot of the reviewers have stopped helping in development).
